ProjectSidewalk/SidewalkWebpage
# Public API v2: Clustered Attributes

Resolves #1241 
Resolves #1242 
Fixes #1247 
Resolves #1317 
Partially fixes #909 
Partially fixes #1184 

Changes include:
* added new public API endpoints to get clustered attributes, along with an endpoint to get the clustered attributes and the labels that make up those clusters
* v2 for the access score endpoints uses the counts of clustered attributes instead of running the old clustering algorithm at the time of the query
* API page updated to reflect the changes to the API in v2
* updated results page to use clustered attributes instead of labels, along with other speed improvements to the results page: on my local machine it now loads in 5 seconds instead of 30 seconds

API page before:
![api-page-before](https://user-images.githubusercontent.com/6518824/46369592-44d3ce80-c638-11e8-81cd-ccb0ea620363.png)

API page after:
![api-page-after](https://user-images.githubusercontent.com/6518824/46369598-48ffec00-c638-11e8-9493-c966d5b9893a.png)

results page before:
![results-choropleth-before](https://user-images.githubusercontent.com/6518824/46369646-68971480-c638-11e8-9ca3-849cbbb97846.png)
![choropleth-counts-before](https://user-images.githubusercontent.com/6518824/46369650-6b920500-c638-11e8-9527-e34dd065a99d.png)

results page after:
![results-choropleth-after](https://user-images.githubusercontent.com/6518824/46369655-6f258c00-c638-11e8-95e8-6dc1289dd776.png)
![choropleth-counts-after](https://user-images.githubusercontent.com/6518824/46369660-7056b900-c638-11e8-8b81-37905468de40.png)

Testing:
1. Start by checking out a specific commit from this branch (9387e763ae22cae49962a0dfb7a05f40a5f5f492). The only commits after this are merge commits. And the stuff from develop that gets merged in has an evolution that truncates the label table, which will make it hard to test that all this stuff works!
1. Check that your attribute (cluster) tables have data in them. If not, go to the /clustering endpoint and run both single and multi user clustering; you will see progress in the terminal where you ran `sbt run` (this will take a long time, possibly over an hour on the full dataset). You should probably run single user clustering then multi user clustering instead of clicking on the button that does both at once since they will take so long to run.
1. Make sure that each of the API endpoints will run and give a reasonable output (might run out of memory for very large queries on the /attributes and /attributesWithLabels endpoints)
1. Check that the results page loads more quickly, and that the label counts are significantly smaller in the new version compared to the old.
1. Just poke around in the v2 API endpoints to see if you can find any holes or anything that seems out of place. Recreate a couple of the queries yourself in SQL, and verify that you get the same results as the API does.
1. Check out the most up to date version of the branch (instead of that older commit). Compile and make sure that everything runs. The label table will be empty, so you may want to go and do some auditing (possibly with a non-admin account?) and increase the lat-lng bounds on an API call to encompass all of DC to make sure you can still get the data from the API. But you won't be able to truly test most of it here. Just make sure that nothing breaks!
